contracts/PooledStaking.sol

```diff
@@ -0,0 +1,512 @@
+
+  function stake(
+    uint amount,
+    address[] calldata _contracts,
+    uint[] calldata _allocations
+  ) onlyMembers external {
+
+    Staker storage staker = stakers[msg.sender];
+
+    require(
+      _contracts.length >= staker.contracts.length,
+      "Allocating to fewer contracts is not allowed"
+    );
+
+    require(
+      _contracts.length == _allocations.length,
+      "Contracts and allocations arrays should have the same length"
+    );
+
+    require(
+      staker.staked > 0,
+      "Allocations can be set only when staked amount is non-zero"
+    );
+
+    Vault.deposit(token, msg.sender, amount);
+    staker.staked = staker.staked.add(amount);
+
+    uint oldLength = staker.contracts.length;
+    uint totalAllocation;
+
+    for (uint i = 0; i < _contracts.length; i++) {
+
+      address contractAddress = _contracts[i];
+      uint oldAllocation = staker.allocations[contractAddress];
+      uint newAllocation = _allocations[i];
+      bool isNewAllocation = i >= oldLength;
+
+      totalAllocation = totalAllocation.add(newAllocation);
+
+      require(newAllocation >= MIN_STAKE, "Allocation minimum not met");
+      require(newAllocation <= staker.staked, "Cannot allocate more than staked");
+
+      if (!isNewAllocation) {
+        require(contractAddress == staker.contracts[i], "Unexpected contract order");
+        require(oldAllocation <= newAllocation, "New allocation is less than previous allocation");
+      }
+
+      if (staker.allocations[contractAddress] == newAllocation) {
+        // no changes to this contract
+        continue;
+      }
+
+      if (isNewAllocation) {
+        staker.contracts.push(contractAddress);
+        contracts[contractAddress].stakers.push(msg.sender);
+      }
+
+      staker.allocations[contractAddress] = newAllocation;
+      contracts[contractAddress].staked = contracts[contractAddress]
+        .staked
+        .sub(oldAllocation)
+        .add(newAllocation);
+    }
+
+    require(
+      totalAllocation <= staker.staked.mul(MAX_LEVERAGE),
+      "Total allocation exceeds maximum allowed"
+    );
+  }
+
+  function unstake(uint amount) onlyMembers external {
+    uint unstakable = getMaxUnstakable(msg.sender);
+    require(unstakable >= amount, "Requested amount exceeds max unstakable amount");
+    stakers[msg.sender].staked = stakers[msg.sender].staked.sub(amount);
+  }
+
+  function requestDeallocation(
+    address[] calldata _contracts,
+    uint[] calldata _amounts,
+    uint[] calldata _insertAfter // deallocation ids to insert deallocation at
+  ) onlyMembers external {
+
+    require(
+      _contracts.length == _amounts.length,
+      "Contracts and amounts arrays should have the same length"
+    );
+
+    require(
+      _contracts.length == _insertAfter.length,
+      "Contracts and _insertAfter arrays should have the same length"
+    );
+
+    Staker storage staker = stakers[msg.sender];
+
+    for (uint i = 0; i < _contracts.length; i++) {
```

Comment:
need to check how much gas these loops use as you increase _contracts and compare with gas block limit

ideally we have a test that monitors gas usage for an acceptable upper limit